Machine Learning Project :

Firstly, we'll ne importing the libraries needed for the project.

In [1]:
import numpy as np
import pandas as pd

Now, we'll be loading the Iris dataset from scikit-learn datasets module.

In [2]:
from sklearn.datasets import load_iris
iris_ds = load_iris()

Here, we'll be checking the keys of iris dataset.

In [3]:
print("Keys of iris_ds: \n{}".format(iris_ds.keys()) )

Keys of iris_ds: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])


Now, we'll be checking what does the key DESCR contain.

In [4]:
print(iris_ds["DESCR"][:200] + "\n...")

Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes
...


Let's check the values of some other keys in similar manner.

In [5]:
print("Target Names : {}".format(iris_ds["target_names"]))
print("Feature Names : \n{}".format(iris_ds["feature_names"]))

Target Names : ['setosa' 'versicolor' 'virginica']
Feature Names : 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


For the project, we'll mainly be using data contained in target and data fields.
Therefore we'll be checking the type and shape of the key data.

In [6]:
print("Type of data : {}".format(type(iris_ds["data"])))
print("Shape of data : {}".format(iris_ds["data"].shape))

Type of data : <class 'numpy.ndarray'>
Shape of data : (150, 4)


Here, we'll be chceking the values the key data comntains.

In [7]:
print("First five columns of data :\n{}".format(iris_ds["data"][:5]))

First five columns of data :
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


Now, we'll be doing the same for the key target.

In [8]:
print("Type of target : {}".format(type(iris_ds["target"])))
print("Shape of target : {}".format(iris_ds["target"].shape))
print("Target ;\n{}".format(iris_ds["target"]))

Type of target : <class 'numpy.ndarray'>
Shape of target : (150,)
Target ;
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Now we'll be splitting the data into train and test for both data(X) and labels(y) using sklearn's train_test_split function. 

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(iris_ds['data'], iris_ds['target'], random_state=0)

Here, we'll check the shape of the four new arrays.

In [10]:
print("X_train shape : {}".format(X_train.shape))
print("X_test shape : {}".format(X_test.shape))
print("y_train shape : {}".format(y_train.shape))
print("y_test shape : {}".format(y_test.shape))

X_train shape : (112, 4)
X_test shape : (38, 4)
y_train shape : (112,)
y_test shape : (38,)


Now we will build the machine learning model. Here, we'll use k-nearest neighbour classifier.

In [11]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)

To build the model on training set, we will fit the model.

In [12]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

Let's put just one data point and test our model.

In [13]:
X_new = np.array([[5, 2.9, 1, 0.2]])
print("X_new.shape: {}".format(X_new.shape))

X_new.shape: (1, 4)


In [14]:
prediction = knn.predict(X_new)
print("Prediction: {}".format(prediction))
print("Predicted target name: {}".format(iris_ds['target_names'][prediction]))

Prediction: [0]
Predicted target name: ['setosa']


Here, we'll evaluate the model and check the accuracy of our model.

In [15]:
y_pred = knn.predict(X_test)
print("Test set predictions:\n {}".format(y_pred))

Test set predictions:
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [16]:
print("Test set score: {}".format(np.mean(y_pred == y_test)))

Test set score: 0.9736842105263158


We can see that our model's accuracy is 97.368%. We conclude that our model is a good fit.